packages/ts/react-grid/test/autoform.spec.tsx

```diff
@@ -0,0 +1,155 @@
+      await assertFormFieldValue(result, 'Some number', person.someNumber);
+    });
+    it('works without an item', async () => {
+      const result = render(<AutoForm service={personService} model={PersonModel} />);
+      await assertFormFieldValue(result, 'First name', '');
+      await assertFormFieldValue(result, 'Last name', '');
+      await assertFormFieldValue(result, 'Email', '');
+      await assertFormFieldValue(result, 'Some number', 0);
+    });
+    it('uses values from an item', async () => {
+      const person = (await getItem(personService, 2))!;
+
+      const result = render(<AutoForm service={personService} model={PersonModel} item={person} />);
+      await assertFormFieldValue(result, 'First name', person.firstName);
+      await assertFormFieldValue(result, 'Last name', person.lastName);
+      await assertFormFieldValue(result, 'Email', person.email);
+      await assertFormFieldValue(result, 'Some number', person.someNumber);
+    });
+    it('submits a valid form', async () => {
+      const service: CrudService<Person> & HasLastFilter = createService<Person>(personData);
+
+      const person = await getItem(service, 1);
+
+      const result = render(<AutoForm service={service} model={PersonModel} item={person} />);
+      await setFormField(result, 'First name', 'foo');
+      await submit(result);
+      await nextFrame();
+      const newItem = await getItem(service, 1);
+      expect(newItem!.firstName).to.equal('foo');
+    });
+
+    it('calls onSubmit with the new item', async () => {
+      const service: CrudService<Person> & HasLastFilter = createService<Person>(personData);
+
+      const person = await getItem(service, 1);
+      let submittedItem: Person | undefined;
+      const result = render(
+        <AutoForm
+          service={service}
+          model={PersonModel}
+          item={person}
+          onSubmit={({ item }) => {
+            submittedItem = item;
+          }}
+        />,
+      );
+      await setFormField(result, 'First name', 'foo');
+      await submit(result);
+
+      await nextFrame();
+      expect(submittedItem!.firstName).to.equal('foo');
+    });
+    it('shows an error if the endpoint call fails', async () => {
+      const service: CrudService<Person> & HasLastFilter = createService<Person>(personData);
+      service.update = async (item: Person): Promise<Person | undefined> => {
+        throw new Error('foobar');
+      };
+      const person = await getItem(service, 1);
+      const result = render(<AutoForm service={service} model={PersonModel} item={person} />);
+      await submit(result);
+      await nextFrame();
+      // eslint-disable-next-line
+      const formError = (result.container.querySelector('#formerror') as HTMLElement).innerText;
+      expect(formError).to.equal('Something went wrong, please check all your values');
+    });
+    it('calls onSubmitEerror and does not show error if the endpoint call fails', async () => {
```

Comment:
Could also test that `onSubmit` is not called when there is an error.

Reply:
Fixed